File: src/main/java/com/baulsupp/oksocial/ws/OkWsCommand.java
package com.baulsupp.oksocial.ws;

import com.baulsupp.oksocial.Main;
import com.baulsupp.oksocial.commands.MainAware;
import com.baulsupp.oksocial.commands.ShellCommand;
import com.google.common.collect.Lists;
import com.baulsupp.oksocial.output.util.UsageException;
import java.util.List;
import java.util.Scanner;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.WebSocket;

public class OkWsCommand implements ShellCommand, MainAware {
  private Main main;

  @Override public void setMain(Main main) {
    this.main = main;
  }

  @Override public String name() {
    return "okws";
  }

  @Override public boolean handlesRequests() {
    return true;
  }

  @Override public List<Request> buildRequests(OkHttpClient client, Request.Builder requestBuilder,
      List<String> arguments) throws Exception {
    if (arguments.size() != 1) {
      throw new UsageException("usage: okws wss://host");
    }

    Request request = main.createRequestBuilder().url(arguments.get(0)).build();

    WebSocketPrinter printer = new WebSocketPrinter(main.outputHandler);
    WebSocket websocket = client.newWebSocket(request, printer);

    Scanner sc = new Scanner(System.in);
    while (sc.hasNextLine()) {
      String line = sc.nextLine();
      websocket.send(line);
    }

    printer.waitForExit();

    return Lists.newArrayList();
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/ws/OkWsCommand.kt
package com.baulsupp.oksocial.ws

import com.baulsupp.oksocial.Main
import com.baulsupp.oksocial.commands.MainAware
import com.baulsupp.oksocial.commands.ShellCommand
import com.baulsupp.oksocial.output.util.UsageException
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.Scanner

class OkWsCommand : ShellCommand, MainAware {
  private var main: Main? = null

  override fun setMain(main: Main) {
    this.main = main
  }

  override fun name(): String {
    return "okws"
  }

  override fun handlesRequests(): Boolean {
    return true
  }

  override fun buildRequests(client: OkHttpClient, arguments: List<String>): List<Request> {
    if (arguments.size != 1) {
      throw UsageException("usage: okws wss://host")
    }

    val request = Request.Builder().url(arguments[0]).build()

    val printer = WebSocketPrinter(main!!.outputHandler!!)
    val websocket = client.newWebSocket(request, printer)

    val sc = Scanner(System.`in`)
    while (sc.hasNextLine()) {
      val line = sc.nextLine()
      websocket.send(line)
    }

    printer.waitForExit()

    return listOf()
  }
}
